### ISSUE TITLE
notebooks/data and train give index out of bounds for axis 2 with size 40

### ISSUE THREAD

**Issue by MarioCavero** [ethz-asl/vgn#23]
I saw [this issue](https://github.com/ethz-asl/vgn/issues/22) opened today. I am having a similar error. I uploaded my own dataset and proceeded with the pipeline. All fine before the clean-up (pasted the notebook code into a `.py` file for my easiness. 

```
python3 notebooks/data.py 
Number of samples: 9960
Number of positives: 1247
Number of negatives: 8713
Traceback (most recent call last):
  File "notebooks/data.py", line 35, in <module>
    tsdf_grid = tsdf.get_grid()
  File "/home/mario/catkin_ws/src/vgn/src/vgn/perception.py", line 114, in get_grid
    grid[0, i, j, k] = distances[idx]
IndexError: index 40 is out of bounds for axis 2 with size 40
```
I changed something (which probably I shouldn't have) in `perception.py` as I read 

> What you can do is, Add the condition to check whether the index value is between [0, 39]

```
    def get_grid(self):
        cloud = self._volume.extract_voxel_point_cloud()
        points = np.asarray(cloud.points)
        distances = np.asarray(cloud.colors)[:, [0]]
        grid = np.zeros((1, 40, 40, 40), dtype=np.float32)
        for idx, point in enumerate(points):
            if(idx<40):
                i, j, k = np.floor(point / self.voxel_size).astype(int)
                grid[0, i, j, k] = distances[idx]
        return grid
```
This allowed the script to work:
```
python3 notebooks/data.py 
Number of samples: 9960
Number of positives: 1247
Number of negatives: 8713

```
But for training now (I ran it with CUDA_LAUNCH_BLOCKING=1 as the error suggested it for better debugging):

```
CUDA_LAUNCH_BLOCKING=1 python scripts/train_vgn.py --dataset data/datasets/foo 
INFO - 2022-07-25 22:52:21,748 - engine - Engine run starting with max_epochs=30.
Epoch [1/30]: [4/280]   1%|5                                       [00:02<03:46]../aten/src/ATen/native/cuda/IndexKernel.cu:91: operator(): block: [0,0,0], thread: [10,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:91: operator(): block: [0,0,0], thread: [30,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
ERROR - 2022-07-25 22:52:25,695 - engine - Current run is terminating due to exception: CUDA error: device-side assert triggered
ERROR - 2022-07-25 22:52:25,719 - engine - Engine run is terminating due to exception: CUDA error: device-side assert triggered
Traceback (most recent call last):
  File "scripts/train_vgn.py", line 216, in <module>
    main(args)
  File "scripts/train_vgn.py", line 86, in main
    trainer.run(train_loader, max_epochs=args.epochs)
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 704, in run
    return self._internal_run()
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 783, in _internal_run
    self._handle_exception(e)
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 466, in _handle_exception
    raise e
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 753, in _internal_run
    time_taken = self._run_once_on_dataset()
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 854, in _run_once_on_dataset
    self._handle_exception(e)
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 466, in _handle_exception
    raise e
  File "/home/mario/anaconda3/envs/vgn/lib/python3.8/site-packages/ignite/engine/engine.py", line 840, in _run_once_on_dataset
    self.state.output = self._process_function(self, self.state.batch)
  File "scripts/train_vgn.py", line 160, in _update
    y_pred = select(net(x), index)
  File "scripts/train_vgn.py", line 119, in select
    label = qual_out[batch_index, :, index[:, 0], index[:, 1], index[:, 2]].squeeze()
RuntimeError: CUDA error: device-side assert triggered
Epoch [1/30]: [4/280]   1%|5                                       [00:02<04:28]

```

I am sure it's related to this index out of bounds but I am not really sure how to fix it. Probably I should add a better workaround in the `perception.py` file for the index condition than the one I already tried.

**Comment by mbreyer** [comment]
The index error you got during training most likely is due to the index of a grasp configuration to be out of bounds. This is different from the error above which happens when building the TSDF grid. Did you also copy the block from the data notebook that removes grasp positions that lie outside the workspace?

**Comment by MarioCavero** [comment]
After some tries and debugging (Sorry for the late reply!), I realised I was running `python scripts/construct_dataset.py data/raw/foo data/datasets/foo` **before** running the clean up method. So the clean up might have been working, but the train was taking the already generating the voxel grids/grasp targets required to train VGN.

For the clean up, I commented out the lines necessary for visualization/rospy, and no problem at all. Train worked fine (although) quicker than expected, probably the amount of grasps got reduced to a certain amount after the clean up!

Thanks!
